ompi/mca/coll/han/coll_han_gatherv.c

```diff
@@ -0,0 +1,381 @@
+                                                    (char *) rbuf
+                                                        + (size_t) displs[w_peer] * rdsize,
+                                                    bounce_buf + offset);
+                offset += rdsize * (size_t) rcounts[w_peer];
+            }
+        }
+
+    root_out:
+        if (low_displs) {
+            free(low_displs);
+        }
+        if (low_rcounts) {
+            free(low_rcounts);
+        }
+        if (up_displs) {
+            free(up_displs);
+        }
+        if (up_rcounts) {
+            free(up_rcounts);
+        }
+        if (up_peer_lb) {
+            free(up_peer_lb);
+        }
+        if (up_peer_ub) {
+            free(up_peer_ub);
+        }
+        if (bounce_buf) {
+            free(bounce_buf);
+        }
+
+        return err;
+    }
+
+    /* #################### Root's local peers ########################### */
+    if (root_up_rank == up_rank) {
+        /* Low Gatherv */
+        low_comm->c_coll->coll_gatherv(sbuf, scount, sdtype, NULL, NULL, NULL, NULL, root_low_rank,
+                                       low_comm, low_comm->c_coll->coll_gatherv_module);
+        return OMPI_SUCCESS;
+    }
+
+    /* #################### Other node followers ########################### */
+    size_t *low_data_size = NULL, sdsize = 0, send_size = 0;
+
+    ompi_datatype_type_size(sdtype, &sdsize);
+    send_size = sdsize * (size_t) scount;
+
+    if (root_low_rank != low_rank) {
+        /* Low Gather - Gather each local peer's send data size */
+        low_comm->c_coll->coll_gather((const void *) &send_size, 1, MPI_UINT64_T, NULL, 1,
+                                      MPI_UINT64_T, root_low_rank, low_comm,
+                                      low_comm->c_coll->coll_gather_module);
+        /* Low Gatherv */
+        low_comm->c_coll->coll_gatherv(sbuf, scount, sdtype, NULL, NULL, NULL, NULL, root_low_rank,
+                                       low_comm, low_comm->c_coll->coll_gatherv_module);
+        return OMPI_SUCCESS;
+    }
+
+    /* #################### Node leaders ########################### */
+
+    char *tmp_buf = NULL;
+    ompi_datatype_t *temptype = MPI_BYTE;
+
+    /* Allocate a temporary array to gather the data size, i.e. data type size x count,
+     * in bytes from local peers */
+    low_data_size = malloc(low_size * sizeof(size_t));
+    if (!low_data_size) {
+        err = OMPI_ERR_OUT_OF_RESOURCE;
+        goto node_leader_out;
+    }
+
+    /* Low Gather -  Gather local peers' send data sizes */
+    low_comm->c_coll->coll_gather((const void *) &send_size, 1, MPI_UINT64_T,
+                                  (void *) low_data_size, 1, MPI_UINT64_T, root_low_rank, low_comm,
+                                  low_comm->c_coll->coll_gather_module);
+
+    /* Determine if we need to create a custom datatype instead of MPI_BYTE,
+     * to avoid count(type int) overflow */
```

Comment:
Can we tag this somehow so it can be removed once we have big count collectives implemented?

Reply:
Yes. I added a TODO.